File: webui/src/main/java/com/taskadapter/webui/pages/NewConfigPage.java
package com.taskadapter.webui.pages;

import com.taskadapter.PluginManager;
import com.taskadapter.config.StorageException;
import com.taskadapter.connector.definition.Descriptor;
import com.taskadapter.web.uiapi.UISyncConfig;
import com.taskadapter.webui.ConfigOperations;
import com.vaadin.ui.Alignment;
import com.vaadin.ui.Button;
import com.vaadin.ui.Component;
import com.vaadin.ui.GridLayout;
import com.vaadin.ui.Label;
import com.vaadin.ui.ListSelect;
import com.vaadin.ui.Panel;
import com.vaadin.ui.TextField;

import java.util.Iterator;

import static com.taskadapter.webui.Page.message;

public class NewConfigPage {

    /**
     * Callback for config creation page.
     */
    public interface Callback {
        /**
         * This method is called after new config was created.
         * 
         * @param config
         *            created config.
         */
        void configCreated(UISyncConfig config);
    }

    private final TextField descriptionTextField;
    private final ListSelect connector1;
    private final ListSelect connector2;
    private final Panel panel;
    private final Label errorMessageLabel;
    private final ConfigOperations configOps;
    private final Callback callback;

    private NewConfigPage(PluginManager pluginManager, ConfigOperations ops,
            Callback callback) {

        this.configOps = ops;
        this.callback = callback;

        panel = new Panel(message("createConfigPage.createNewConfig"));
        panel.setWidth("600px");

        final GridLayout grid = new GridLayout(2, 4);
        grid.setSpacing(true);
        grid.setMargin(true);
        panel.setContent(grid);

        connector1 = createSystemListSelector(message("createConfigPage.system1"), pluginManager);
        grid.addComponent(connector1, 0, 0);

        connector2 = createSystemListSelector(message("createConfigPage.system2"), pluginManager);
        grid.addComponent(connector2, 1, 0);

        descriptionTextField = new TextField(message("createConfigPage.description"));
        descriptionTextField.setInputPrompt(message("createConfigPage.optional"));
        descriptionTextField.setWidth("100%");
        grid.addComponent(descriptionTextField, 0, 1, 1, 1);
        grid.setComponentAlignment(descriptionTextField, Alignment.MIDDLE_CENTER);

        // empty label by default
        errorMessageLabel = new Label();
        errorMessageLabel.addStyleName("error-message-label");

        final Button saveButton = new Button(message("createConfigPage.create"));
        saveButton.addClickListener(new Button.ClickListener() {
            @Override
            public void buttonClick(Button.ClickEvent event) {
                saveClicked();
            }
        });

        grid.addComponent(errorMessageLabel, 0, 2, 1, 2);
        grid.addComponent(saveButton, 1, 3);
        grid.setComponentAlignment(saveButton, Alignment.MIDDLE_RIGHT);
    }

    private static ListSelect createSystemListSelector(String title, PluginManager plugins) {
        final ListSelect res = new ListSelect(title);
        res.setRequired(true);
        res.setNullSelectionAllowed(false);

        final Iterator<Descriptor> connectors = plugins.getPluginDescriptors();
        while (connectors.hasNext()) {
            final Descriptor connector = connectors.next();
            res.addItem(connector.getID());
            res.setItemCaption(connector.getID(), connector.getLabel());
        }

        res.setRows(res.size());
        return res;
    }

    private void saveClicked() {
        try {
            validate();
            errorMessageLabel.setValue("");
            callback.configCreated(save());
        } catch (ConnectorNotSelectedException e) {
            errorMessageLabel.setValue(e.getMessage());
        } catch (StorageException e) {
            errorMessageLabel.setValue(message("createConfigPage.failedToSave"));
        }
    }

    private void validate() throws ConnectorNotSelectedException {
        if (connector1.getValue() == null) {
            connector1.setRequiredError(message("createConfigPage.pleaseSelectSystem1"));
            throw new ConnectorNotSelectedException(message("createConfigPage.pleaseSelectSystem1"));
        } else {
            connector1.setRequiredError("");
        }

        if (connector2.getValue() == null) {
            connector1.setRequiredError(message("createConfigPage.pleaseSelectSystem2"));
            throw new ConnectorNotSelectedException(message("createConfigPage.pleaseSelectSystem2"));
        } else {
            connector2.setRequiredError("");
        }
    }

    private UISyncConfig save() throws StorageException {

        final String descriptionString = descriptionTextField.getValue();
        final String id1 = (String) connector1.getValue();
        final String id2 = (String) connector2.getValue();

        return configOps.createNewConfig(descriptionString, id1, id2);
    }

    private class ConnectorNotSelectedException extends Exception {

        private static final long serialVersionUID = 1L;

        public ConnectorNotSelectedException(String string) {
            super(string);
        }
    }

    /**
     * null Renders a page for the new configs.
     * 
     * @param pluginManager
     *            plugin manager.
     * @param ops
     *            configuration operations.
     * @param callback
     *            callback for the item.
     * @return UI component.
     */
    public static Component render(PluginManager pluginManager,
            ConfigOperations ops, Callback callback) {
        return new NewConfigPage(pluginManager, ops, callback).panel;
    }
}

File: webui/src/main/scala/com/taskadapter/webui/pages/NewConfigPage.scala
package com.taskadapter.webui.pages

import com.taskadapter.config.StorageException
import com.taskadapter.web.service.Sandbox
import com.taskadapter.web.uiapi.{ConfigId, SetupId}
import com.taskadapter.webui.{BasePage, ConfigOperations, EventTracker, Layout, SessionController}
import com.taskadapter.webui.Page.message
import com.taskadapter.webui.service.Preservices
import com.vaadin.flow.component.Text
import com.vaadin.flow.component.dependency.CssImport
import com.vaadin.flow.component.html.Label
import com.vaadin.flow.router.Route

trait Callback {
  /**
    * This method is called after new config was created.
    *
    * @param configId unique id of the new config.
    */
  def configCreated(configId: ConfigId): Unit
}

@Route(value = Navigator.NEW_CONFIG, layout = classOf[Layout])
@CssImport(value = "./styles/views/mytheme.css")
class NewConfigPage extends BasePage {

  private val configOps: ConfigOperations = SessionController.buildConfigOperations()
  private val services: Preservices = SessionController.getServices
  private val sandbox: Sandbox = SessionController.createSandbox()
  private val pluginManager = services.pluginManager
  private val editorManager = services.editorManager

  val wizard = new WizardPanel()

  var connector1Id: Option[String] = None
  var connector1SetupId: Option[SetupId] = None
  var connector2Id: Option[String] = None
  var connector2SetupId: Option[SetupId] = None
  var description: Option[String] = None

  wizard.registerStep(new SelectConnectorWizardStep(pluginManager, next =
    (connectorId) => {
      connector1Id = Some(connectorId)
      wizard.showStep(2)
    }
  ))

  wizard.registerStep(new NewConfigConfigureSystem(editorManager, configOps, sandbox, setupSelected =
    (label) => {
      connector1SetupId = Some(label)
      wizard.showStep(3)
    }
  ))

  wizard.registerStep(new SelectConnectorWizardStep(pluginManager, next =
    (connectorId) => {
      connector2Id = Some(connectorId)
      wizard.showStep(4)
    }
  ))

  wizard.registerStep(new NewConfigConfigureSystem(editorManager, configOps, sandbox, setupSelected =
    (label) => {
      connector2SetupId = Some(label)
      wizard.showStep(5)
    }
  ))

  wizard.registerStep(new NewConfigGiveDescription(d => {
    description = Some(d)
    saveClicked()
  }
  ))

  wizard.showStep(1)

  // empty label by default
  val errorMessageLabel = new Text("")
//  errorMessageLabel.addClassName("error-message-label")

  add(new Label(message("createConfigPage.createNewConfig")),
    errorMessageLabel,
    wizard.ui)

  private def saveClicked(): Unit = {
    try {
      val configId = save()
      Navigator.configsList()
    } catch {
      case e: StorageException =>
        errorMessageLabel.setText(message("createConfigPage.failedToSave"))
    }
  }

  @throws[StorageException]
  private def save(): ConfigId = {
    val descriptionString = description.get
    val id1 = connector1Id.get
    val id2 = connector2Id.get
    configOps.createNewConfig(descriptionString, id1, connector1SetupId.get, id2, connector2SetupId.get)
  }

}
